distributed/diagnostics/progress.py

```diff
@@ -291,3 +294,80 @@ def transition(self, key, start, finish, *args, **kwargs):
+
+        # Time bin size. TODO: make this configurable?
+        self.dt = 1.0
+        now = time.time()
+
+        # Timestamps for tracking compute durations by task group.
+        # Start with length 2 so that we always can compute a valid dt later.
+        self.time: list[float] = [now] * 2
+        # The amount of compute since the last timestamp
+        self.compute: dict[str, list[float]] = {}
+        # The number of threads at the time
+        self.nthreads: list[float] = [scheduler.total_nthreads] * 2
+
+    def transition(self, key, start, finish, *args, **kwargs):
+        # We are mostly interested in when tasks complete for now, so just look
+        # for when processing transitions to memory. Later we could also extend
+        # this if we can come up with useful visual channels to show it in.
+        if start == "processing" and finish == "memory":
+            startstops = kwargs.get("startstops")
+            if not startstops:
+                logger.warn(
+                    f"Task {key} finished processing, but timing information seems to be missing"
+                )
+                return
+
+            # Possibly extend the timeseries if another dt has passed
+            now = time.time()
+            self.time[-1] = now
+            while self.time[-1] - self.time[-2] > self.dt:
+                self.time[-1] = self.time[-2] + self.dt
+                self.time.append(now)
+                self.nthreads.append(self.scheduler.total_nthreads)
+                for g in self.compute.values():
+                    g.append(0.0)
+
+            # Get the task
+            task = self.scheduler.tasks[key]
+            group = task.group
+
+            # If the group is new, add it to the timeseries as if it has been
+            # here the whole time
+            if group.name not in self.compute:
+                self.compute[group.name] = [0.0] * len(self.time)
+
+            for startstop in startstops:
+                if startstop["action"] != "compute":
+                    continue
+                stop = startstop["stop"]
+                start = startstop["start"]
+                idx = len(self.time) - 1
+                # If the stop time is after the most recent bin,
+                # roll back the current index. Not clear how often this happens.
+                while idx > 0 and self.time[idx - 1] > stop:
+                    idx -= 1
+                # Allocate the timing information of the task to the time bins.
+                while idx > 0 and stop > start:
+                    delta = stop - max(self.time[idx - 1], start)
+                    self.compute[group.name][idx] += delta
+
+                    stop -= delta
+                    idx -= 1
+
+    def restart(self, scheduler):
+        now = time.time()
+        self.time = [now] * 2  # make sure to respect the requirement that len > 2
+        self.nthreads = [self.scheduler.total_nthreads] * 2
+        self.compute = {}
```

Comment:
I would probably encapsulate this into an `_init` method since this is the same you call in `__init__`. wrongfully resetting state is a very likely source for bugs

Reply:
:+1: